src/lib/codap-helper.ts

```diff
@@ -152,6 +162,100 @@ export class CodapHelper {
+    const dataContext = await this.getDataContext(existingDataContextName);
+    if (dataContext) {
+      // first run through both DCs and gather the attribute details for each
+      const originalAttributes = [] as Array<{name: string; collection: string; index: number; attr: any }>;
+      const sharedAttributes = [] as Array<{name: string; collection: string; index: number; attr: any; }>;
+
+      dataContext.collections.forEach(collection => {
+        collection.attrs.forEach((attr, i) => {
+          originalAttributes.push({name: attr.name, collection: collection.name, index: i, attr});
+        });
+      });
+
+      sharedDataContext.collections.forEach(collection => {
+        collection.attrs.forEach((attr, i) => {
+          sharedAttributes.push({name: attr.name, collection: collection.name, index: i, attr});
+        });
+      });
+
+      // If we have any new collections, create them in our existing DC without attributes
+      const newCollections = sharedDataContext.collections.filter(collA => {
+        return !dataContext.collections.some(collB => collA.name === collB.name);
+      });
+
+      if (newCollections.length > 0) {
+        newCollections.forEach(collection => collection.attrs = []);
+        await this.addCollections(existingDataContextName, newCollections);
+      }
+
+      // then create any new attributes as necessary
+      const newAttributes = sharedAttributes.filter(attrA => {
+        return !originalAttributes.some(attrB => attrA.name === attrB.name);
+      });
+
+      if (newAttributes.length > 0) {
+        const attributeCreationCommands: any[] = [];
+        // list of unique collections the new attributes belong to
+        const collectionsForNewAttributes = Array.from(new Set(newAttributes.map(a => a.collection)));
+        collectionsForNewAttributes.forEach(collectionName => {
+          // group the new attributes by collection and create a command to create all the new
+          // attributes for that collection
+          const newAttributesInCollection = newAttributes
+            .filter(a => a.collection === collectionName)
+            .map(a => a.attr);
+          attributeCreationCommands.push({
+            action: "create",
+            resource: dataContextResource(existingDataContextName, `collection[${collectionName}].attribute`),
+            values: newAttributesInCollection
+          });
+        });
+        // create all new attributes at once
+        await codapInterface.sendRequest(attributeCreationCommands);
+      }
+
+      // then move any existing attributes as necessary
+      const movedAttributes = originalAttributes.filter(attrA => {
+        const attrB = sharedAttributes.find(a => a.name === attrA.name);
+        return attrB && (attrA.collection !== attrB.collection || attrA.index !== attrB.index);
+      });
+
+      if (movedAttributes.length > 0) {
+        const moveCreationCommands = movedAttributes.map(attr => {
```

Comment:
Perhaps `attributeMoveCommands` to be analogous to `attributeCreationCommands`.

Reply:
That's what it was in my head!

Reply:
I had a feeling that might be the case. 😉